service/history/replication/raw_task_converter.go

```diff
@@ -712,3 +637,323 @@ func convertGetHistoryError(
+	var versionHistoryLength int32
+	var versionHistoryItem *historyspb.VersionHistoryItem
+
+	currentVersionHistory, err := versionhistory.GetCurrentVersionHistory(c.mutableState.GetExecutionInfo().VersionHistories)
+	if err != nil {
+		return nil, err
+	}
+	for _, versionHistory := range c.progress.eventVersionHistoryItems {
+		item, err := versionhistory.FindLCAVersionHistoryItemFromItemSlice(versionHistory, currentVersionHistory.Items)
+		if err != nil {
+			return nil, err
+		}
+
+		if versionHistoryItem == nil ||
+			// if seeing LCA item with higher event ID
+			item.GetEventId() > versionHistoryItem.GetEventId() ||
+			// if seeing LCA item with equal event ID but shorter history
+			(item.GetEventId() == versionHistoryItem.GetEventId() && int32(len(versionHistory)) < versionHistoryLength) {
+			versionHistoryLength = int32(len(versionHistory))
+			versionHistoryItem = item
+		}
+	}
+	return versionHistoryItem, nil
+}
+
+func (c *syncVersionedTransitionTaskConverter) generateSyncStateMutationFromTask(
+	transitionHistory []*persistencespb.VersionedTransition,
+) (*replicationspb.ReplicationTask, error) {
+	events, next_page_token, err := c.getEvents(
+		transitionHistory,
+		c.taskInfo.FirstEventID,
+		c.taskInfo.VersionedTransition.NamespaceFailoverVersion,
+		true,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &replicationspb.ReplicationTask{
+		TaskType:     enumsspb.REPLICATION_TASK_TYPE_SYNC_VERSIONED_TRANSITION_TASK,
+		SourceTaskId: c.taskInfo.TaskID,
+		Priority:     c.taskInfo.Priority,
+		Attributes: &replicationspb.ReplicationTask_SyncVersionedTransitionTaskAttributes{
+			SyncVersionedTransitionTaskAttributes: &replicationspb.SyncVersionedTransitionTaskAttributes{
+				EventBatches:  events,
+				NextPageToken: next_page_token,
+				// TODO: SyncStateMutationAttributes
+			},
+		},
+		VersionedTransition: transitionHistory[len(transitionHistory)-1],
+		VisibilityTime:      timestamppb.New(c.taskInfo.VisibilityTimestamp),
+	}, nil
+}
+
+func (c *syncVersionedTransitionTaskConverter) generateSyncStateMutationFromCache(
+	transitionHistory []*persistencespb.VersionedTransition,
+	lca *historyspb.VersionHistoryItem,
+) (*replicationspb.ReplicationTask, error) {
+	events, next_page_token, err := c.getEvents(
+		transitionHistory,
+		lca.GetEventId(),
+		lca.GetVersion(),
+		false,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &replicationspb.ReplicationTask{
+		TaskType:     enumsspb.REPLICATION_TASK_TYPE_SYNC_VERSIONED_TRANSITION_TASK,
+		SourceTaskId: c.taskInfo.TaskID,
+		Priority:     c.taskInfo.Priority,
+		Attributes: &replicationspb.ReplicationTask_SyncVersionedTransitionTaskAttributes{
+			SyncVersionedTransitionTaskAttributes: &replicationspb.SyncVersionedTransitionTaskAttributes{
+				EventBatches:  events,
+				NextPageToken: next_page_token,
+				// TODO: SyncStateMutationAttributes
```

Comment:
I think we can do this and also snapshot in a separate PR. I'd imaging a lot of logic from https://github.com/temporalio/temporal/pull/6381 can be reused here.

Reply:
will solve this in another PR combining with the SyncState API.